code/Model/SiteTree.php

```diff
@@ -1338,51 +1338,103 @@ public function collateDescendants($condition, &$collator)
-     * @param bool $includeTitle Show default <title>-tag, set to false for custom templating
-     * @return string The XHTML metatags
+     * @see HTML::createTag()
+     * @return array
      */
-    public function MetaTags($includeTitle = true)
+    public function MetaComponents()
     {
-        $tags = array();
-        if ($includeTitle && strtolower($includeTitle) != 'false') {
-            $tags[] = HTML::createTag('title', array(), $this->obj('Title')->forTemplate());
-        }
+        $tags = [];
+
+        $tags['title'] = [
+            'tag' => 'title',
+            'attributes' => [],
+            'content' => $this->obj('Title')->forTemplate()
+        ];
 
         $generator = trim(Config::inst()->get(self::class, 'meta_generator'));
         if (!empty($generator)) {
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'generator',
-                'content' => $generator,
-            ));
+            $tags['generator'] = [
+                'attributes' => [
+                    'name' => 'generator',
+                    'content' => $generator,
+                ],
+            ];
         }
 
         $charset = ContentNegotiator::config()->uninherited('encoding');
-        $tags[] = HTML::createTag('meta', array(
-            'http-equiv' => 'Content-Type',
-            'content' => 'text/html; charset=' . $charset,
-        ));
+        $tags['contentType'] = [
+            'attributes' => [
+                'http-equiv' => 'Content-Type',
+                'content' => 'text/html; charset=' . $charset,
+            ],
+        ];
         if ($this->MetaDescription) {
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'description',
-                'content' => $this->MetaDescription,
-            ));
+            $tags['description'] = [
+                'attributes' => [
+                    'name' => 'description',
+                    'content' => $this->MetaDescription,
+                ],
+            ];
         }
 
         if (Permission::check('CMS_ACCESS_CMSMain')
             && $this->ID > 0
         ) {
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'x-page-id',
-                'content' => $this->obj('ID')->forTemplate(),
-            ));
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'x-cms-edit-link',
-                'content' => $this->obj('CMSEditLink')->forTemplate(),
-            ));
+            $tags['pageId'] = [
+                'attributes' => [
+                    'name' => 'x-page-id',
+                    'content' => $this->ID,
+                ],
+            ];
+            $tags['cmsEditLink'] = [
+                'attributes' => [
+                    'name' => 'x-cms-edit-link',
+                    'content' => $this->CMSEditLink(),
+                ],
+            ];
+        }
+
+        $this->extend('MetaComponents', $tags);
+
+        return $tags;
```

Comment:
If $tags was an ArrayList instead of Array, we could use `<% loop $MetaComponents %>` in templates couldn't we?

You know, I just realised we really need a `casting` type for lists as well so we don't have to auto-wrap these in arraylists.

```
private static $casting = [
    'MetaComponents' => ArrayList::class,
];
```

Note: This doesn't work, I just wish it did. :P

Reply:
> You know, I just realised we really need a casting type for lists as well so we don't have to auto-wrap these in arraylists.

That would be cool :)

> If $tags was an ArrayList instead of Array, we could use <% loop $MetaComponents %> in templates couldn't we?

The MetaTags are a bit of a mixed bag. We could use a template but the tag type is variable (due to inclusion of title tag 🙄), as are the attributes. So it would need to look something like:

```html
<$TagType <% if $Attributes %><% loop $Attributes %>$Name="$Value.ATT"<% end_loop %><% end_if %><% if $Contents %>>$Contents</$TagType><% else %> /><% end_if %>
```

Note that `$MetaTags` would need to render that template to retain backwards compatibility.

I don't think there's a big advantage in switching to a template at this point, unless it's with the view of removing `$MetaTags` in 5 and working only with `$MetaComponents` instead.

I'm not a big fan of changing  more than necessary where backwards compatibility is a priority. So far this PR isn't changing the way the data is rendered, I'm just exposing that data before it is flattened, so it should be nice and safe. Switching to templates probably wouldn't hurt but the more you change the more risk of side effects.

BTW I thought about splitting out the title tag from MetaComponents so it would only provide true `meta` tags, but the expectation is that `$MetaTags` can give you a title and I want `MetaComponents()` to be the source for everything in `$MetaTags`, so I decided to keep it all together. It makes the structure of the MetaComponents array more complicated than I'd like but I've tried to address that with clear documentation in the method and proposed addition to the Docs.